app/driver.go

```diff
@@ -109,12 +109,35 @@ func deployDriver(c *cli.Context) error {
 		return errors.Wrap(err, "unable to get k8s client")
 	}
 
+	driverDetected := false
 	driver := c.String(FlagDriver)
 	if driver == "" {
+		driverDetected = true
 		driver, err = chooseDriver(kubeClient)
+		logrus.Debugf("Driver %s will be used after automatic detection", driver)
 		if err != nil {
 			return err
 		}
+	} else {
+		logrus.Debugf("Explicitly choose the %s driver", driver)
+	}
+
+	if driver == FlagDriverCSI {
+		csiChecked, err := csi.CheckMountPropagationWithPodSpec(kubeClient, managerImage, os.Getenv(types.EnvPodNamespace))
+		if err != nil {
+			logrus.Warnf("Got an error when checking MountPropagation with pod spec, %v", err)
+		}
+
+		// if user explicitly choose CSI but we cannot deploy.
+		// In this case we should error out instead.
+		if !csiChecked {
+			if driverDetected {
+				logrus.Infof("MountPropagation check failed, should fall back to the flexvolume")
+				driver = FlagDriverFlexvolume
+			} else {
+				return errors.New("CSI cannot be deployed because MountPropagation is not set on kubelet and api-server")
```

Comment:
Use `fmt.Errorf` instead. It's an item checked by go report (https://goreportcard.com/report/github.com/rancher/longhorn-manager).